dcbtool.8
.\" LICENSE
.\"
.\" This software program is released under the terms of a license agreement between you ('Licensee') and Intel.  Do not use or load this software or any associated materials (collectively, the 'Software') until you have carefully read the full terms and conditions of the LICENSE located in this software package.  By loading or using the Software, you agree to the terms of this Agreement.  If you do not agree with the terms of this Agreement, do not install or use the Software.
.\"
.\" * Other names and brands may be claimed as the property of others.
.\"
.TH dcbtool 8 "June 29, 2009"
.SH NAME
dcbtool \- manage the Data Center Bridging (DCB) settings of a DCB interface
.SH SYNOPSIS
.B dcbtool -h
.br
.B dcbtool -v
.br
.B dcbtool [-rR]
.br
.BI "dcbtool [-rR] [" "command" "] [" "command arguments" "]"

.SH DESCRIPTION
.B dcbtool
is used to query and set the DCB settings of a
DCB capable Ethernet interface.  It connects to the client interface
of
.B lldpad
to perform these operations.
.B dcbtool
will operate in interactive mode if it is executed without a 
.I command.
In interactive mode,
.B dcbtool
also functions as an event listener and will print out events
received from
.B lldpad
as they arrive.
.SH OPTIONS
.TP
.B \-h
show the
.B dcbtool
usage message
.TP
.B \-v
shows
.B dcbtool
version information
.TP
.B \-r
display the raw
.B lldpad
client interface messages as well as the readable output.
.TP
.B \-R
display only the raw
.B lldpad
client interface messages
.SH COMMANDS
.TP
.B help 
shows the dcbtool usage message
.TP
.B ping 
test command.
.B lldpad
responds with "PPONG" if the client interface is operational.
.TP
.B license
displays
.B dcbtool
license information
.TP
.B quit
exit from interactive mode

.PP
The following commands interact with
.B lldpad
to manage the daemon and DCB features on DCB capable interfaces.

.B dcbtool
general configuration commands:
.TP
.RB "<" "gc" "|" "go" ">" " dcbx"
get the configured or operational version of the DCB capabilities exchange
protocol.  If different, the configured version will take effect (and become
the operational version) after
.B lldpad
is restarted.
.TP
.BR "sc dcbx v:" "[" "cin" "|" "cee" "|" "force-cin" "|" "force-cee" "]"
set the version of the DCB capabilities exchange protocol which will be used
the next time
.B lldpad
is started.  Information about the CIN version can be found at:
.br
<http://download.intel.com/technology/eedc/dcb_cep_spec.pdf>
Information about the CEE version can be found at:
.br
<http://www.ieee802.org/1/files/public/docs2008/az-wadekar-dcbx-capability-exchange-discovery-protocol-1108-v1.01.pdf>
.br
Note that the dcbx setting is a global setting and changes only take effect
when
.B lldpad
is restarted.  The default DCBX version used is the IEEE standard version.
If a pre-IEEE DCBX version is detected (per port), then
.B lldpad
will fall back to the configured global dcbx setting.  If the dcbx setting
is set to either 'force-cin' or 'force-cee' then any port doing DCBX will
start out in the corresponding legacy DCBX mode.
.PP
DCB per-interface commands:
.TP
\fBgc \fR<\fIifname\fR> <\fIfeature\fR>
get configuration of \fIfeature\fR on interface \fIifname\fR.
.TP
\fBgo \fR<\fIifname\fR> <\fIfeature\fR>
get operational status of \fIfeature\fR on interface \fIifname\fR.
.TP
\fBgp \fR<\fIifname\fR> <\fIfeature\fR>
get peer configuration of \fIfeature\fR on interface \fIifname\fR.
.TP
\fBsc \fR<\fIifname\fR> <\fIfeature\fR> <\fIargs\fR>
set the configuration of \fIfeature\fR on interface \fIifname\fR.
.PP
.I feature
may be one of the following:
.TP
.B dcb
DCB state of the port
.TP
.B pg
priority groups
.TP
.B pfc
priority flow control
.TP
\fBapp:\fR<\fIsubtype\fR>
application specific data
.TP
\fBll:\fR<\fIsubtype\fR>
logical link status
.PP
.I subtype
can be:
.TP
.BR "0" "|" "fcoe"
Fiber Channel over Ethernet (FCoE)
.PP
.I args
can include:
.TP
.BR "e:" "<" "0" "|" "1" ">"
controls feature enable
.TP
.BR "a:" "<" "0" "|" "1" ">"
controls whether the feature is advertised via DCBX to the peer
.TP
.BR "w:" "<" "0" "|" "1" ">"
controls whether the feature is willing to change its operational
configuration based on what is received from the peer
.TP
.RI "[" "feature specific args" "]"
arguments specific to a DCB feature

.PP
Feature specific arguments for
.BR "dcb" ":"
.TP
.BR "on" "|" "off"
enable or disable DCB for the interface.  The
.B go
and
.B gp
commands are not needed for the
.B dcb
feature.  Also, the
.RI "enable(" "e" "), advertise(" "a" ") and willing(" "w" ")"
arguments are not required.

.PP
Feature specific arguments for
.BR "pg" ":"
.TP
.BI "pgid:" "xxxxxxxx"
Priority group ID for the 8 priorities.  From left to right
(priorities 0-7),
.I x
is the corresponding priority group ID value, which
can be
.I 0-7
for priority groups with bandwidth allocations or
.I f
(priority group ID 15) for the unrestricted priority group.

.TP
.BI "pgpct:" "x" "," "x" "," "x" "," "x" "," "x" "," "x" "," "x" "," "x"
Priority group percentage of link bandwidth.  From left to right
(priority groups 0-7),
.I x
is the percentage of link bandwidth allocated
to the corresponding priority group.  The total bandwidth must equal 100%.

.TP
.BI "uppct:" "x" "," "x" "," "x" "," "x" "," "x" "," "x" "," "x" "," "x"
Priority percentage of priority group bandwidth.  From left to right
(priorities 0-7),
.I x
is the percentage of priority group bandwidth
allocated to the corresponding priority.  The sum of percentages for
priorities which belong to the same priority group must total 100% (except
for priority group 15).

.TP
.BI "strict:" "xxxxxxxx"
Strict priority setting.  From left to right (priorities 0-7),
.I x
.RB "is " "0" " or " "1" ".  " "1"
indicates that the priority may utilize all of the bandwidth allocated to
its priority group.

.TP
.BI "up2tc:" "xxxxxxxx"
Priority to traffic class mapping.  From left to right (priorities 0-7),
.I x
is the traffic class (0-7) to which the priority is mapped.

.PP
Feature specific arguments for
.BR "pfc" ":"
.TP
.BI "pfcup:" "xxxxxxxx"
Enable/disable priority flow control.  From left to right
(priorities 0-7),
.I x
.RB "is " "0" " or " "1" ".  " "1"
indicates that the corresponding priority is configured to transmit
priority pause.

.PP
Feature specific arguments for \fBapp:\fR<\fIsubtype\fR>:
.TP
.BI "appcfg:" "xx"
.I xx
is a hexadecimal value representing an 8 bit bitmap where
1 bits indicate the priorities which frames for the applications specified by
.I subtype
should use.  The lowest order bit maps to priority 0.

.PP
Feature specific arguments for \fBll:\fR<\fIsubtype\fR>:
.TP
\fBstatus:\fR[\fB0\fR|\fB1\fR]
For testing purposes, the logical link status may be set to 0 or 1.  This
setting is not persisted in the configuration file.

.SH EXAMPLES

.PP
Enable DCB on interface \fIeth2\fR
.PP
.B dcbtool sc eth2 dcb on

.PP
Assign priorites 0-3 to priority group 0, priorites 4-6 to priority group 1 and
priority 7 to the unrestricted priority.  Also, allocate 25% of link
bandwidth to priority group 0 and 75% to group 1.
.PP
.B dcbtool sc eth2 pg pgid:0000111f pgpct:25,75,0,0,0,0,0,0 

.PP
Enable transmit of Priority Flow Control for priority 3 and assign FCoE to
priority 3.
.PP
.B dcbtool sc eth2 pfc pfcup:00010000
.br
.B dcbtool sc eth2 app:0 appcfg:08

.SH COPYRIGHT
dcbtool - DCB configuration utility
.br
Copyright(c) 2007-2010 Intel Corporation.
.BR
Portions of dcbtool are based on:
.IP hostapd-0.5.7
.IP Copyright 
(c) 2004-2008, Jouni Malinen <j@w1.fi>

.SH LICENSE
This program is free software; you can redistribute it and/or modify it
under the terms and conditions of the GNU General Public License,
version 2, as published by the Free Software Foundation.
.LP
This program is distributed in the hope it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.
.LP
You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
.LP
The full GNU General Public License is included in this distribution in
the file called "COPYING".
.SH SUPPORT
Contact Information:
e1000-eedc Mailing List <e1000-eedc@lists.sourceforge.net>
Intel Corporation, 5200 N.E. Elam Young Parkway, Hillsboro, OR 97124-6497



ChangeLog
Changes from 0.9.39 through 0.940
- Add iSCSI TLV support
Changes from 0.9.26 through 0.9.39
- Numerous bug fixes and code cleanup
- Reworked netlink event handling to be more robust and handle dropped nlmsgs.
- DCBx puts netdevice into linkmode until dcbx negotiation complete
Changes from 0.9.21 through 0.9.26
- Fork dcbd into lldpad and make the daemon a generic LLDP agent deamon.
Changes from 0.9.19 through 0.9.21
- Update libconfig to version 1.3.2.  Fix code to match API changes.
- Add dcbd -k -s options in README and manpage.
- Fix dcbd seg fault when dcbd.conf version string is not 1.0.
- Move the dcbd.conf out of the /etc/sysconfig/ directory into /var/lib/dcbd.
Changes from 0.9.15 through 0.9.19
- Send App priority information to driver via netlink (if kernel supports)
- dcbtool returns non-zero status on error
- Allow user priority percentages (uppct) to be changed while using DCBX v2.
- Improved dcbtool parameter parsing to detect invalid arguments.
- Fixed segfault which occurred in some cases of the config file was removed.
- Fixed dcbtool get app oper error.
Changes from 0.9.7 through 0.9.15
- Set FeatureSeqNo to 1 in the case of peer expiration.
- Use the MSG_DONTWAIT flag in the recv socket call to avoid potential of making
  a blocking call - which would hang dcbd.
- Move initialization of control interface before call to init ports to
  avoid use of uninitialized variable.
- Add support of SAN MAC address, get SAN MAC address through driver interface. 
- Add -s, -k command line options.
- Dcbd, dcbtool man pages and README are updated.
- Set the SO_PRIORITY of the LLDP socket to TC_PRIO_CONTROL (7).  This will
  help the driver to know not to priority tag the LLDP packets.
- Add shared memory segment to save some DCBX state information when
  dcbd stops.  Used to restore DCBX state on restart of dcbd.
- Change default dcbx version to 2.
- Only send the version 2 DCBX TLV when in dcbx version 2 mode (instead of
  both version TLVs).
- Fix dcbd init script so that it loads the dcbnl module for kernels which
  have the option of a dcbnl module.
- A dcbtool error message was changed from "Device not found" to 
  "Device not found, link down or DCB not enabled"
- Each DCB feature (PFC, PG, ...) was modified in driver to indicate whether
  or not a hw reset is required when the feature configuration changes.
  If a HW reset will occur, then dcbd will not react to the next link bounce.
- Ensure that HW flow control settings are synced with the DCBX operational
  state.
- Runlevel 4 removed from the dcbd init script.
- Fix error flag for willing==willing case.
- Don't print out error message anymore in remove_adapter for the default
  config object for objects which aren't created for the default config.
- Fixed potential memory leak in rare shutdown scenarios
- Deleted line of unreachable dead code.
- Remove BCN support from dcbd and dcbtool
- Limit values for DcbxSubtype  - Write Operational DcbxSubtype  for query
- For dcbx version 2, rebalance bandwidth % assignements on set of Priority
  Group feature


dcbtool.c
/*******************************************************************************

  LLDP Agent Daemon (LLDPAD) Software 
  Copyright(c) 2007-2010 Intel Corporation.

  Substantially modified from:
  hostapd-0.5.7
  Copyright (c) 2002-2007, Jouni Malinen <jkmaline@cc.hut.fi> and
  contributors

  This program is free software; you can redistribute it and/or modify it
  under the terms and conditions of the GNU General Public License,
  version 2, as published by the Free Software Foundation.

  This program is distributed in the hope it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
  more details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc.,
  51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.

  The full GNU General Public License is included in this distribution in
  the file called "COPYING".

  Contact Information:
  e1000-eedc Mailing List <e1000-eedc@lists.sourceforge.net>
  Intel Corporation, 5200 N.E. Elam Young Parkway, Hillsboro, OR 97124-6497

*******************************************************************************/

#include <stdlib.h>
#include <stdio.h>
#include <signal.h>
#include <dirent.h>
#include <unistd.h>

#include "clif.h"
#include "lldp_dcbx_cmds.h"
#include "clif_msgs.h"
#include "lldpad.h"
#include "dcbtool.h"
#include "version.h"

static int show_raw;

static const char *cli_version =
"dcbtool v" DCBTOOL_VERSION "\n"
"Copyright (c) 2007-2010, Intel Corporation\n"
"\nSubstantially modified from:  hostapd_cli v 0.5.7\n"
"Copyright (c) 2004-2007, Jouni Malinen <j@w1.fi> and contributors";


static const char *cli_license =
"This program is free software. You can distribute it and/or modify it\n"
"under the terms of the GNU General Public License version 2.\n"
"\n";
/*
"Alternatively, this software may be distributed under the terms of the\n"
"BSD license. See README and COPYING for more details.\n";
*/

static const char *cli_full_license =
"This program is free software; you can redistribute it and/or modify\n"
"it under the terms of the GNU General Public License version 2 as\n"
"published by the Free Software Foundation.\n"
"\n"
"This program is distributed in the hope that it will be useful,\n"
"but WITHOUT ANY WARRANTY; without even the implied warranty of\n"
"MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"
"GNU General Public License for more details.\n"
"\n"
"You should have received a copy of the GNU General Public License\n"
"along with this program; if not, write to the Free Software\n"
"Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA\n"
"\n"
"Alternatively, this software may be distributed under the terms of the\n"
"BSD license.\n"
"\n"
"Redistribution and use in source and binary forms, with or without\n"
"modification, are permitted provided that the following conditions are\n"
"met:\n"
"\n"
"1. Redistributions of source code must retain the above copyright\n"
"   notice, this list of conditions and the following disclaimer.\n"
"\n"
"2. Redistributions in binary form must reproduce the above copyright\n"
"   notice, this list of conditions and the following disclaimer in the\n"
"   documentation and/or other materials provided with the distribution.\n"
"\n"
"3. Neither the name(s) of the above-listed copyright holder(s) nor the\n"
"   names of its contributors may be used to endorse or promote products\n"
"   derived from this software without specific prior written permission.\n"
"\n"
"THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS\n"
"\"AS IS\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT\n"
"LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR\n"
"A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT\n"
"OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,\n"
"SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT\n"
"LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,\n"
"DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY\n"
"THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT\n"
"(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE\n"
"OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.\n"
"\n";

static const char *commands_help =
"DCBX Commands:\n"
"  <gc|go> dcbx                 get configured or operational DCBX versions\n"
"                               the configured version takes effect on next\n"
"                               restart of the lldpad service\n"
"  sc dcbx v:cee                set the DCBX version to be used after next\n"
"          v:cin                lldpad restart.  Version can be set to:\n"
"          v:force-cin          cin: fall back to CIN DCBX if non-IEEE peer\n"
"          v:force-cee          cee: fall back to CEE DCBX if non-IEEE peer\n"
"                               force-cin, force-cee: do not start with IEEE DCBX\n"
"Per Port Commands:\n"
"  gc <ifname> <feature>        get configuration of <feature> on port <ifname>\n"
"  go <ifname> <feature>        get operational status of <feature>\n"
"                               on port <ifname>\n"
"  gp <ifname> <feature>        get peer configuration of <feature>\n"
"                               on port <ifname>\n"
"  sc <ifname> <feature> <args> general form of set feature configuration\n\n"
"  'feature' can be:\n"
"     dcb                       DCB state of port\n"
"     pg                        priority groups\n"
"     pfc                       priority flow control\n"
"     app:<subtype>             application specific data\n"
"     ll:<subtype>              logical link status\n\n"
"  'subtype' can be:\n"
"     [0|fcoe]                  FCoE\n\n"
"     [1|iscsi]                 iSCSI\n\n"
"  'args' can include:\n"
"     [e:<0|1>]                 controls feature enable\n"
"     [a:<0|1>]                 controls feature advertise via DCBX\n"
"     [w:<0|1>]                 controls feature DCBX willing mode\n"
"     [feature specific args]   arguments specific to a feature\n\n"
"  feature specific arguments for 'dcb':\n"
"     [on|off]                  the 'dcb' configuration does not use the\n"
"                               enable, advertise or willing mode parameters\n"
"                               'dcb' also does not use the 'go' or 'gp'\n"
"                               commands.\n\n"
"  feature specific arguments for 'pg':\n"
"     [pgid:xxxxxxxx]           priority group ID of user priority.\n"
"                               From left to right (priorities 0-7), x is\n"
"                               the corresponding priority group value\n"
"                               0-7 (for priority groups with bandwidth)\n"
"                               or 'f' (the no bandwidth allocation group)\n"
"     [pgpct:x,x,x,x,x,x,x,x]   priority group percent of link.\n"
"                               From left to right (priority groups 0-7),\n"
"                               x is the percentage of link bandwidth\n"
"                               assigned. The total must equal 100%.\n"
"     [uppct:x,x,x,x,x,x,x,x]   user priority percent of bandwidth group.\n"
"                               From left to right (priorities 0-7),\n"
"                               x is the percentage of priority group\n"
"                               bandwidth assigned to the priority.\n"
"                               The sum of percentages for priorities in the\n"
"                               same priority group must be 100.\n"
"     [strict:xxxxxxxx]         strict priority setting.\n"
"                               From left to right (priorities 0-7),\n"
"                               x is 0 or 1.  1 indicates that the priority\n"
"                               may utilize all of the bandwidth allocated\n"
"                               to its priority group.\n"
"     [up2tc:xxxxxxxx]          user priority to traffic class mapping.\n"
"                               From left to right (priorities 0-7), x is\n"
"                               a the corresponding traffic class (0-7)\n"
"                               (this argument is currently ignored).\n\n"
"  feature specific arguments for 'pfc':\n"
"     [pfcup:xxxxxxxx]          enable/disable priority flow control.\n"
"                               From left to right (priorities 0-7),\n"
"                               x is 0 or 1.  1 indicates that the priority\n"
"                               is configured to transmit priority pause.\n\n"
"  feature specific arguments for 'app:<subtype>':\n"
"     [appcfg:xx]               'xx' is a hexadecimal value representing an\n"
"                               8 bit bitmap where 1 bits indicate the\n"
"                               priority which frames for the application\n"
"                               specified by the subtype should use.\n"
"                               The lowest order bit maps to priority 0.\n\n"
"  feature specific arguments for 'll:<subtype>':\n"
"     [status:<0|1>]            for testing, the logical link status may\n"
"                               be set to 0 or 1.  This setting is not\n"
"                               persisted.\n\n"
"  help                         show command information\n"
"  license                      show license information\n";

static struct clif *clif_conn;
static int cli_quit = 0;
static int cli_attached = 0;
static const char *clif_iface_dir = CLIF_IFACE_DIR;
static char *clif_ifname = NULL;


static void usage(void)
{
	fprintf(stderr, "%s\n", cli_version);
	fprintf(stderr, 
		"\n"	
		"Usage:\n"
		"  dcbtool -h\n"
		"  dcbtool -v\n"
		"  dcbtool [-rR] [ command ]\n"
		"  dcbtool                      interactive mode\n"
		"\n"
		"Options:\n"
		"  -h                           help (show this usage text)\n"
		"  -v                           shown version information\n"
		"  -r                           show raw messages\n"
		"  -R                           show only raw messages\n"
		"%s",
		commands_help);
}

void print_raw_message(char *msg, int print)
{
	if (!print || !(print & SHOW_RAW))
		return;

	if (!(print & SHOW_RAW_ONLY)) {
		switch (msg[0]) {
		case EVENT_MSG:
			printf("event: ");
			break;
		case CMD_RESPONSE:
			printf("rsp: ");
			break;
		default:
			printf("cmd: ");
			break;
		}
	}
	printf("%s\n", msg);
}

int parse_print_message(char *msg, int print)
{
	int status = 0;

	status = parse_response(msg);

	print_raw_message(msg, print);

	if (print & SHOW_RAW_ONLY)
		return status;

	if (msg[0] == CMD_RESPONSE)
		print_response(msg, status);
	else if (msg[0] == EVENT_MSG)
		print_event_msg(msg);

	return status;
}

static struct clif *cli_open_connection(const char *ifname)
{
	char *cfile;
	int flen;

	if (ifname == NULL)
		return NULL;

	flen = strlen(clif_iface_dir) + strlen(ifname) + 2;
	cfile = malloc(flen);
	if (cfile == NULL)
		return NULL;
	snprintf(cfile, flen, "%s/%s", clif_iface_dir, ifname);

	clif_conn = clif_open(cfile);
	free(cfile);
	return clif_conn;
}


static void cli_close_connection(void)
{
	if (clif_conn == NULL)
		return;

	if (cli_attached) {
		clif_detach(clif_conn);
		cli_attached = 0;
	}
	clif_close(clif_conn);
	clif_conn = NULL;
}


static void cli_msg_cb(char *msg, size_t len)
{
	parse_print_message(msg, SHOW_OUTPUT | show_raw);
}


/* structure of the print argument bitmap:
 *     SHOW_NO_OUTPUT (0x0) - don't print anything for the command
 *     SHOW_OUTPUT (0x01)   - print output for the command
 *     SHOW_RAW (0x02)      - print the raw clif command messages
 *     SHOW_RAW_ONLY (0x04) - print only the raw clif command messages
*/
static int _clif_command(struct clif *clif, char *cmd, int print)
{
	char buf[MAX_CLIF_MSGBUF];
	size_t len;
	int ret;

	print_raw_message(cmd, print);

	if (clif_conn == NULL) {
		printf("Not connected to lldpad - command dropped.\n");
		return -1;
	}
	len = sizeof(buf) - 1;
	ret = clif_request(clif, cmd, strlen(cmd), buf, &len,
			       cli_msg_cb);
	if (ret == -2) {
		printf("'%s' command timed out.\n", cmd);
		return -2;
	} else if (ret < 0) {
		printf("'%s' command failed.\n", cmd);
		return -1;
	}
	if (print) {
		buf[len] = '\0';
		ret = parse_print_message(buf, print);
	}

	return ret;
}


inline int clif_command(struct clif *clif, char *cmd, int raw)
{
	return _clif_command(clif, cmd, SHOW_OUTPUT | raw);
}


static int cli_cmd_ping(struct clif *clif, int argc, char *argv[], int raw)
{
	return clif_command(clif, "P", raw);
}

static int cli_cmd_help(struct clif *clif, int argc, char *argv[], int raw)
{
	printf("%s", commands_help);
	return 0;
}


static int cli_cmd_license(struct clif *clif, int argc, char *argv[], int raw)
{
	printf("%s\n\n%s\n", cli_version, cli_full_license);
	return 0;
}


static int cli_cmd_quit(struct clif *clif, int argc, char *argv[], int raw)
{
	cli_quit = 1;
	return 0;
}


static void cli_list_interfaces(struct clif *clif)
{
	struct dirent *dent;
	DIR *dir;

	dir = opendir(clif_iface_dir);
	if (dir == NULL) {
		printf("Control interface directory '%s' could not be "
		       "opened.\n", clif_iface_dir);
		return;
	}

	printf("Available interfaces:\n");
	while ((dent = readdir(dir))) {
		if (strcmp(dent->d_name, ".") == 0 ||
		    strcmp(dent->d_name, "..") == 0)
			continue;
		printf("%s\n", dent->d_name);
	}
	closedir(dir);
}


static int cli_cmd_interface(struct clif *clif, int argc,
				     char *argv[], int raw)
{
	if (argc < 1) {
		cli_list_interfaces(clif);
		return 0;
	}

	cli_close_connection();
	free(clif_ifname);
	clif_ifname = strdup(argv[0]);

	if (cli_open_connection(clif_ifname)) {
		printf("Connected to interface '%s.\n", clif_ifname);
		if (clif_attach(clif_conn, NULL) == 0)
			cli_attached = 1;
		else
			printf("Warning: Failed to attach to lldpad.\n");
	} else {
		printf(
		     "Could not connect to lldpad interface '%s' - re-trying\n",
			clif_ifname);
	}
	return 0;
}


struct cli_cmd {
	const char *cmd;
	int (*handler)(struct clif *clif, int argc, char *argv[], int raw);
};

static struct cli_cmd cli_commands[] = {
	{ "ping", cli_cmd_ping },
	{ "help", cli_cmd_help },
	{ "interface", cli_cmd_interface },
	{ "license", cli_cmd_license },
	{ "quit", cli_cmd_quit },
	{ NULL, NULL }
};


static int request(struct clif *clif, int argc, char *argv[], int raw)
{
	struct cli_cmd *cmd, *match = NULL;
	int count;
	int ret	= 0;

	count = 0;
	cmd = cli_commands;
	while (cmd->cmd) {
		if (strncasecmp(cmd->cmd, argv[0], strlen(argv[0])) == 0) {
			match = cmd;
			count++;
		}
		cmd++;
	}

	if (count > 1) {
		printf("Ambiguous command '%s'; possible commands:", argv[0]);
		cmd = cli_commands;
		while (cmd->cmd) {
			if (strncasecmp(cmd->cmd, argv[0], strlen(argv[0])) ==
			    0) {
				printf(" %s", cmd->cmd);
			}
			cmd++;
		}
		printf("\n");
		ret = -1;
	} else if (count == 0) {
		ret = handle_dcb_cmds(clif, argc, &argv[0], raw);
	} else {
		ret = match->handler(clif, argc - 1, &argv[1], raw);
	}

	return ret;	
}


static void cli_recv_pending(struct clif *clif, int in_read)
{
	int first = 1;
	if (clif_conn == NULL)
		return;
	while (clif_pending(clif)) {
		char buf[256];
		size_t len = sizeof(buf) - 1;
		if (clif_recv(clif, buf, &len) == 0) {
			buf[len] = '\0';
			if (in_read && first)
				printf("\n");
			first = 0;
			cli_msg_cb(buf, len);
		} else {
			printf("Could not read pending message.\n");
			break;
		}
	}
}


static void cli_interactive(int raw)
{
	const int max_args = 10;
	char cmd[256], *res, *argv[max_args], *pos;
	int argc;

	printf("\nInteractive mode\n\n");

	do {
		cli_recv_pending(clif_conn, 0);
		printf("> ");
		alarm(1);
		res = fgets(cmd, sizeof(cmd), stdin);
		alarm(0);
		if (res == NULL)
			break;
		pos = cmd;
		while (*pos != '\0') {
			if (*pos == '\n') {
				*pos = '\0';
				break;
			}
			pos++;
		}
		argc = 0;
		pos = cmd;
		for (;;) {
			while (*pos == ' ')
				pos++;
			if (*pos == '\0')
				break;
			argv[argc] = pos;
			argc++;
			if (argc == max_args)
				break;
			while (*pos != '\0' && *pos != ' ')
				pos++;
			if (*pos == ' ')
				*pos++ = '\0';
		}
		if (argc)
			request(clif_conn, argc, argv, raw);
	} while (!cli_quit);
}


static void cli_terminate(int sig)
{
	cli_close_connection();
	exit(0);
}


static void cli_alarm(int sig)
{
	if (clif_conn && _clif_command(clif_conn, "P", SHOW_NO_OUTPUT)) {
		printf("Connection to lldpad lost - trying to reconnect\n");
		cli_close_connection();
	}
	if (!clif_conn) {
		clif_conn = cli_open_connection(clif_ifname);
		if (clif_conn) {
			printf("Connection to lldpad re-established\n");
			if (clif_attach(clif_conn, NULL) == 0)
				cli_attached = 1;
			else
				printf(
				      "Warning: Failed to attach to lldpad.\n");
		}
	}
	if (clif_conn)
		cli_recv_pending(clif_conn, 1);
	alarm(1);
}


int main(int argc, char *argv[])
{
	int interactive;
	int raw = 0;
	int warning_displayed = 0;
	int c;
	int ret = 0;

	for (;;) {
		c = getopt(argc, argv, "hvrR");
		if (c < 0)
			break;
		switch (c) {
		case 'h':
			usage();
			return 0;
		case 'r':
			if (raw) {
				usage();
				return -1;
			}
			raw = SHOW_RAW;
			break;
		case 'R':
			if (raw) {
				usage();
				return -1;
			}
			raw = (SHOW_RAW | SHOW_RAW_ONLY);
			break;
		case 'v':
			printf("%s\n", cli_version);
			return 0;
		default:
			usage();
			return -1;
		}
	}
	show_raw = raw;

	interactive = argc == optind;

	if (interactive) {
		printf("%s\n\n%s\n\n", cli_version,
		       cli_license);
	}

	for (;;) {
		if (clif_ifname == NULL) {
			struct dirent *dent;
			DIR *dir = opendir(clif_iface_dir);
			if (dir) {
				while ((dent = readdir(dir))) {
					if (strcmp(dent->d_name, ".") == 0 ||
					    strcmp(dent->d_name, "..") == 0)
						continue;
					clif_ifname = strdup(dent->d_name);
					break;
				}
				closedir(dir);
			}
		}
		clif_conn = cli_open_connection(clif_ifname);
		if (clif_conn) {
			if (warning_displayed)
				printf("Connection established.\n");
			break;
		}

		if (!interactive) {
			perror("Failed to connect to lldpad - clif_open");
			return -1;
		}

		if (!warning_displayed) {
			printf("Could not connect to lldpad - re-trying\n");
			warning_displayed = 1;
		}
		sleep(1);
		continue;
	}

	signal(SIGINT, cli_terminate);
	signal(SIGTERM, cli_terminate);
	signal(SIGALRM, cli_alarm);

	if (interactive) {
		if (clif_attach(clif_conn, NULL) == 0)
			cli_attached = 1;
		else
			printf("Warning: Failed to attach to lldpad.\n");
		cli_interactive(raw);
	} else
		ret = request(clif_conn, argc - optind, &argv[optind], raw);

	free(clif_ifname);
	cli_close_connection();
	return ret;
}
